### PR TITLE
sql: de-objectify planNode

### PR DESCRIPTION
**Pull request by knz**
This series of 3 commits is part of a larger strategy: make the "logical plan" part of planNode dumb and method-less; and migrate the "execution" part of planNode to a (tbd) separate interface that only implements `Start`, `Next`, `Values` and `Close`, possibly living in a separate package.

The only remaining methods that are not part of execution here are `MarkDebug` and `DebugValues` - #16514

From the discussion with Jordan:

> it serves to continue the spreading out of related functionality all around the codebase, a style that I personally find hard to follow.

The key here is to understand what the unit of functionality/reasoning.

For local *execution* the unit of reasoning is the iterator (from the Volcano model) and thus it is useful to reason "locally" per iterator/node and see the Start, Next and Values method side-by-side, one set per file and one file per node type. This is not changing here.

For *query transformations* however the unit of reasoning is not the node but rather the transformation algorithm. 
Radu and I can both tell you than when operating on a logical plan we reason about combinations of nodes, not individual nodes any more. Also grouping the code together allows us to recognize more readily when the transformation for different node types is actually doing the same thing. With the code as it now stands thanks to this patch, we reveal for example a potential gain by embedding a common struct that has the `ordering` and `columns`. The amount of boilerplate is also reduced, since a `switch` allows us to define a default behavior.

This specific section of the IR RFC deals with this point: https://github.com/knz/cockroach/blob/589ae688a446157318144a2800d8b47d63b89425/docs/RFCS/sql-ir.md#general-principles-behind-the-approach

### CODE REVIEW

**Comment by knz** [comment]
review plz

I will rebase afterwards

**Comment by jordanlewis** [comment]
The change seems sound... assuming that this is the correct choice for overall code organization and readability. I'm not quite convinced either way! Both seem to have their advantages. This new approach hints that the methods are pure functions and don't change their operands, which is quite nice. On the other hand, it serves to continue the spreading out of related functionality all around the codebase, a style that I personally find hard to follow. To me, having to jump between at least 3 different files (columns, spans, and ordering) to figure out how a single plan node behaves is frustrating, to say the least. I don't think I have sufficient operational memory in my brain to keep all of these disparate implementation pieces in my head at once - to me, it makes a lot of sense to have them all nearby in a single file.

Also, even though you've done a great job documenting where these new pseudo-methods live at the top of `plan.go`, I'm afraid that this sort of style will raise even further the barrier of entry to this code for a newcomer. The standard tools of "examine what interface methods are available for this kind of data" will not work, since the implementation lives in several functions that are file-wise distant from the data.

I suppose this is just a bit of a classic OOP vs the world battle. Which provides cleaner, more maintainable and readable code overall - polymorphism with implementations close to the data they act on but far from each other, or switch-case behavior with implementations far from the data they act on but close by to each other? Perhaps my preference for OOP is showing through too strongly here.

On the positive side of things, your new organization shows just how clean and empty most of the implementations are, which is ... calming! And I of course trust your good judgement - you have a much better sense of the overall structure and flow of how all of this code works together than I do.

Just throwing all this out there for discussion - I'm not trying to throw a wrench into getting this good progress merged.

---

Reviewed 34 of 34 files at r1, 35 of 35 files at r2, 44 of 44 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan_columns.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUF-yPTfkGgxRYdeN-:-KmUF-yPTfkGgxRYdeN0:b-g8vkxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1999328e3165116ef656a542c0a0ace72828b0a1/pkg/sql/plan_columns.go#L79)):*
> ```Go
> 
> 		// Nodes using the RETURNING helper.
> 	case *deleteNode:
> ```

coalesce cases?

---

*[pkg/sql/plan_columns.go, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUEykxiJzk4QcZl3t4:-KmUEykyl_x-z1i3Dlr4:b-7jkyjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1999328e3165116ef656a542c0a0ace72828b0a1/pkg/sql/plan_columns.go#L99)):*
> ```Go
> 		return planColumns(n.left)
> 
> 	}
> ```

panic on unmatched case?

---

*[pkg/sql/plan_ordering.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUE2cfHIdgjB4jdUBq:-KmUE2cgFoPTBqAdiW6y:b1b8epd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L64)):*
> ```Go
> 	case *renderNode:
> 		return n.ordering
> 	case *sortNode:
> ```

Pull this out into a function maybe? you've done that in the previous file.

---

*[pkg/sql/plan_ordering.go, line 115 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUDqeaPeCzQz2OFaWB:-KmUDqebJRtZChBjnzpk:b-26kkpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L115)):*
> ```Go
> 		}
> 		return ord
> 	}
> ```

Do we want to panic if we have an uncovered case here like in the other file?

---

*[pkg/sql/plan_spans.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUBmUAil4v9DQbHdac:-KmUBmUAil4v9DQbHdad:b5wqhot) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> //
> // Author: Nathan VanBenschoten <nathan@cockroachlabs.com>
> ```

s/nathan/knz/?

---

*[pkg/sql/plan_spans.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUCBrKXTSebWZlBo3A:-KmUCBrKXTSebWZlBo3B:bjmzykp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L36)):*
> ```Go
> 	switch n := plan.(type) {
> 	case *valueGenerator:
> 		return nil, nil, nil
> ```

collapse these three cases?

---

*[pkg/sql/plan_spans.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUCE_le9YPkFFq2ATT:-KmUBz5vvxLW4n92AsAX:b5jywhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L53)):*
> ```Go
> 
> 	case *delayedNode:
> 		return collectSpans(ctx, n.plan)
> ```

it's a shame that these cases can't all be collapsed! i see why they can't be - just saying.

---

*[pkg/sql/plan_spans.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUBsCa29d2Xq7dibyE:-KmUBsCa29d2Xq7dibyF:bwc1181) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L87)):*
> ```Go
> 	}
> 
> 	panic("unimplemented")
> ```

put the node type into the panic message so it's easier to figure out what's going on if this line gets hit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-:-KmUF6CjatNCricU14gz:b-z8g0tx)*

**Comment by andreimatei** [comment]
LGTM

Given that these methods don't mutate anything and that they seem quasi-related to compilation, I had similar questions to Jordan. But I'd down with the new centralized methods, I think the code is cleaner ¯\_(ツ)_/¯

---

Reviewed 7 of 34 files at r1, 1 of 35 files at r2, 1 of 44 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan_columns.go, line 22 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmXO8Yca7XqLoYZduu9:-KmXO8Yca7XqLoYZduuA:b-xtj2nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1999328e3165116ef656a542c0a0ace72828b0a1/pkg/sql/plan_columns.go#L22)):*
> ```Go
> var noColumns = make(sqlbase.ResultColumns, 0)
> 
> // planColumns returns the column names and types. The length of the
> ```

please expand this comment a bit. What are "the col names and types"? Say something about how each node is a data source, and tie this method more clearly to Values() and Next() somehow.

---

*[pkg/sql/plan_ordering.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUE2cfHIdgjB4jdUBq:-KmXNkxjKKYYHXXhaXJ1:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L64)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Pull this out into a function maybe? you've done that in the previous file.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-:-KmXXBJwGyAtmU9qJ8MU:b-e2c60w)*

**Comment by knz** [comment]
Thank you for the thoughtful comments!

Jordan I agree with you that having to jump from one file to another when trying to comprehend a logical unit of functionality/engineering is irritating and counter-productive.

The key here is to understand what the unit of functionality/reasoning.

For local *execution* the unit of reasoning is the iterator (from the Volcano model) and thus it is useful to reason "locally" per iterator/node and see the Start, Next and Values method side-by-side, one set per file and one file per node type. I do not plan on changing this here.

For *query transformations* however the unit of reasoning is not the node but rather the transformation algorithm. 
Radu and I can both tell you than when operating on a logical plan we reason about combinations of nodes, not individual nodes any more. Also grouping the code together allows us to recognize more readily when the transformation for different node types is actually doing the same thing. With the code as it now stands thanks to this patch, we reveal for example a potential gain by embedding a common struct that has the `ordering` and `columns`. The amount of boilerplate is also reduced, since a `switch` allows us to define a default behavior.

I encourage you to read this specific section of the IR RFC which deals with exactly this point: https://github.com/knz/cockroach/blob/589ae688a446157318144a2800d8b47d63b89425/docs/RFCS/sql-ir.md#general-principles-behind-the-approach

---

Review status: 0 of 47 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan_columns.go, line 22 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmXO8Yca7XqLoYZduu9:-KmXcYyBo9td1UWF-Bh4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1999328e3165116ef656a542c0a0ace72828b0a1/pkg/sql/plan_columns.go#L22)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please expand this comment a bit. What are "the col names and types"? Say something about how each node is a data source, and tie this method more clearly to Values() and Next() somehow.
</blockquote></details>

Done.

---

*[pkg/sql/plan_columns.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUF-yPTfkGgxRYdeN-:-KmXYMOYvWytzSzoAqOT:b-pnw9qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1999328e3165116ef656a542c0a0ace72828b0a1/pkg/sql/plan_columns.go#L79)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

coalesce cases?
</blockquote></details>

This is Go. Its type system does not allow it.

I will make another pass later at this code overall (incl. the other functions) so that nodes that have equivalent attributes (e.g. `rh`, `columns`, `ordering`) embed a common struct and we can use a single type switch.

---

*[pkg/sql/plan_columns.go, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUEykxiJzk4QcZl3t4:-KmXYglC3KrLc217Olhz:b-xac47u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1999328e3165116ef656a542c0a0ace72828b0a1/pkg/sql/plan_columns.go#L99)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

panic on unmatched case?
</blockquote></details>

That would be incorrect - for now the logical plan nodes have "results" and thus asking for their column set is a valid operation. 
(Incidentally, EXPLAIN uses this.)

---

*[pkg/sql/plan_ordering.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUE2cfHIdgjB4jdUBq:-KmXbSZkxQGckhX1rGIK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L64)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[pkg/sql/plan_ordering.go, line 115 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUDqeaPeCzQz2OFaWB:-KmXaQm3KydZttgxNLO0:b-4wjurz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L115)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Do we want to panic if we have an uncovered case here like in the other file?
</blockquote></details>

No - the default is intentional in fact -- all the other nodes simply provide no ordering guarantee on their results.

---

*[pkg/sql/plan_spans.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUBmUAil4v9DQbHdac:-KmXaY6jqdKCM1V9_yaW:b-66wmzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L15)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/nathan/knz/?
</blockquote></details>

Nathan wrote the code. I merely grouped it together.

---

*[pkg/sql/plan_spans.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUCBrKXTSebWZlBo3A:-KmXaqBh3XyJjfXHpDs-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L36)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

collapse these three cases?
</blockquote></details>

Done.

---

*[pkg/sql/plan_spans.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUCE_le9YPkFFq2ATT:-KmXaqvwr-G4LvnW2Zr3:bmls417) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L53)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

it's a shame that these cases can't all be collapsed! i see why they can't be - just saying.
</blockquote></details>

Go's type system = :poop: 

---

*[pkg/sql/plan_spans.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUBsCa29d2Xq7dibyE:-KmXawIraZr0AVnYF5zW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7cfaf238553c2de64c042406a39dfe1c25ca1e/pkg/sql/plan_spans.go#L87)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

put the node type into the panic message so it's easier to figure out what's going on if this line gets hit
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-:-KmXcgMjLrSjMvq6Me_0:b1bmg4j)*

**Comment by knz** [comment]
Rebased; PTAL

- needed for #16514

**Comment by andreimatei** [comment]
:lgtm:

---

Reviewed 1 of 47 files at r4, 1 of 35 files at r5, 4 of 44 files at r6.
Review status: 6 of 47 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/plan_columns.go, line 24 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmgB8qYzyyTE2g9NjVe:-KmgB8qYzyyTE2g9NjVf:bqsphb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b27d6731af59e0dab39c8284cdea000c01d622bc/pkg/sql/plan_columns.go#L24)):*
> ```Go
> // planColumns returns the signature of rows logically computed
> // by the given planNode.
> // The signature consist's of the list of columns with
> ```

consists

---

*[pkg/sql/plan_ordering.go, line 115 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUDqeaPeCzQz2OFaWB:-KmgBVyOUoC8D87LdHyv:bcske1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L115)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No - the default is intentional in fact -- all the other nodes simply provide no ordering guarantee on their results.
</blockquote></details>

consider adding a comment about it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-:-KmgBzd17uJUQhb_Gi66:bnfp4nl)*

**Comment by knz** [comment]
Review status: 6 of 47 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/plan_columns.go, line 24 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmgB8qYzyyTE2g9NjVe:-KmgSAX-pq8DGig4Om8a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b27d6731af59e0dab39c8284cdea000c01d622bc/pkg/sql/plan_columns.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consists
</blockquote></details>

Done.

---

*[pkg/sql/plan_ordering.go, line 115 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-KmUDqeaPeCzQz2OFaWB:-KmgSBXiCTMFlV_M2skG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a350f3f8aa3fdfb40d5e689e64c6a92513cb0a/pkg/sql/plan_ordering.go#L115)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider adding a comment about it
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241)*

**Comment by jordanlewis** [comment]
:lgtm:

---

Reviewed 47 of 47 files at r4.
Review status: 2 of 47 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16241#-:-KmgSsEnrarFjUsy-PRZ:bnfp4nl)*

**Comment by knz** [comment]
TFYRs! :heart_decoration:
